java/arcs/core/analysis/AccessPathLabels.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+
+/**
+ * An abstract value that keeps track of [InformationFlowLabels] for different [AccessPath] values.
+ *
+ * This abstract value uses a map from [AccessPath] to [InformationFlowLabels] to keep track of the
+ * labels for each access path. We use a reduced bottom construction, i.e., whenever a value for an
+ * access path is `BOTTOM`, the whole map is considered `BOTTOM`.
+ */
+data class AccessPathLabels private constructor(
+    private val _accessPathLabels: BoundedAbstractElement<Map<AccessPath, InformationFlowLabels>>
+) : AbstractValue<AccessPathLabels> {
+    override val isBottom = _accessPathLabels.isBottom
+    override val isTop = _accessPathLabels.isTop
+
+    val accessPathLabels: Map<AccessPath, InformationFlowLabels>?
+        get() = _accessPathLabels.value
+
+    override infix fun isEquivalentTo(other: AccessPathLabels) =
+        _accessPathLabels.isEquivalentTo(other._accessPathLabels) { a, b ->
+            if (a.size == b.size) a.all { (accessPath, aLabels) ->
+                b[accessPath]?.isEquivalentTo(aLabels) ?: false
+            } else false
```

Comment:
Optional, but I think adding braces here for the if- and else-statement would make this more readable.

Reply:
Actually, this could just be:

```
a.size == b.size && a.all {
  ...
}
```

Reply:
Good idea!